### PR TITLE
Project 1 Proposal: Vim Syntax Highlighting for Bril

### LINKED ISSUE #27: Project 1 Proposal: Vim Syntax Highlighting for Bril
**What will you do?**
Implement syntax highlighting for Bril.

**How will you do it?**
I'll learn how to make a vim plugin using existing syntax highlighters as reference, and any guides I might find online.

**How will you empirically measure success?**
I'll test the highlighting on files covering all language features that are meant to be highlighted.

### PR DESCRIPTION
**Pull request by epeguero**
closes #27

### CODE REVIEW

File: content/blog/2019-09-25-vim-syntax-highlighting.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+This is useful, since it allows us to factor out and reuse regexes from syntax group definitions into meta-strings. 
+However, Vimscript since makes a distinction between *strings* and *literal strings*, reasoning about "meta-strings" and "meta-literal-strings" can lead to hard to detect bugs.
+
+Semantically, a Vimscript string consists of a list of characters and characters 'escaped' by a backslash. (hence "\" is an invalid string; it is a malformed character escape).
+On the other hand, a literal string simply consists of the literal characters in the string.
+If we attempt to factor out a regexp as a string, any escaped characters will be stored as 'escaped', rather than as characters.
+Splicing this faulty regexp into a program string will generate a faulty program string containing escaped characters.
+
+We can solve this by *escaping each 'escaping' backslash with a backslash, except for those backslashes that escape backlashes.* (it's no wonder the leading Vimscript tutorial recommends [beer](http://learnvimscriptthehardway.stevelosh.com/chapters/08.html) to accompany some of the exercises.)
```
Comment (sampsyo) [inline content/blog/2019-09-25-vim-syntax-highlighting.md]:
```suggestion
We can solve this by *escaping each 'escaping' backslash with a backslash, except for those backslashes that escape backlashes.* (It's no wonder the leading Vimscript tutorial recommends [beer](http://learnvimscriptthehardway.stevelosh.com/chapters/08.html) to accompany some of the exercises.)
```

Comment (sampsyo) [reply content/blog/2019-09-25-vim-syntax-highlighting.md]: :beers:

File: content/blog/2019-09-25-vim-syntax-highlighting.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+To avoid backslash hell, we can instead store regexps as literal strings, which can never escape characters.
+Later on, these can be spliced with program fragments that are also represented as string literals, rather than strings, to avoid escaping any hard-coded regexps.
+
+Once I dealt with the frustrations and mysterious bugs, I found Vimscript to be pretty fun, in part due to its flexibility,
+and I enjoyed pondering the feedback problem of language design for language tooling.
+
+### Design Considerations and Future Work
+As I've described it, a syntax highlighters principle job is to give a meaningful visual presentation to an evolving syntax tree.
```
Comment (sampsyo) [inline content/blog/2019-09-25-vim-syntax-highlighting.md]:
```suggestion
As I've described it, a syntax highlighter's principal job is to give a meaningful visual presentation to an evolving syntax tree.
```

File: content/blog/2019-09-25-vim-syntax-highlighting.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+In reality, developers have come to expect much more from a syntax highlighter, including:
+- Syntactically-aware code editing and navigation, such as automatic indentation, folding, variable renaming, etc...
```
Comment (sampsyo) [inline content/blog/2019-09-25-vim-syntax-highlighting.md]:
```suggestion
- Syntactically-aware code editing and navigation, such as automatic indentation, folding, variable renaming, etc.
```

I think "etc." is fine by itself; it doesn't need to be doubled up with an ellipsis.

File: content/blog/2019-09-25-vim-syntax-highlighting.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+- Type checking and type information lookup
+- Code linting
+- etc..
```
Comment (sampsyo) [inline content/blog/2019-09-25-vim-syntax-highlighting.md]:
```suggestion
- etc.
```

**Review by sampsyo — CHANGES_REQUESTED** [review]
This is awesome! I really like the ruminations on the spiritual path of a syntax highlighter. Here are just a few low-level writing suggestions.

**Review by epeguero — COMMENTED** [review]
Thanks for the edits.
I made a change to that grammatically weird sentence, clarified what I meant a bit.
Let me know if there are issues with my commits; I'm learning how to resolve a pull request review as I go.

**Comment by sampsyo** [comment]
Nice work! This looks great. Merged & published!

**Pull request merged**